8272215: Add InetAddress methods for parsing IP address literals

AlekseiEfimov opened the pull request:
### Summary 

The changes in this PR add new API to `java.net.InetAddress`, `java.net.Inet4Address`, and
 `java.net.Inet6Address` classes to parse IP address literals:
 ```
method public static java.net.InetAddress java.net.InetAddress.ofLiteral(java.lang.String)
method public static java.net.Inet4Address java.net.Inet4Address.ofLiteral(java.lang.String)
method public static java.net.InetAddress java.net.Inet6Address.ofLiteral(java.lang.String)
``` 

### How new methods differ from existing ones

These methods differ from `InetAddress.getByName` and `InetAddress.getAllByName` in the following ways:
1. If a string supplied is not an address literal it is not forwarded to the system-wide resolver, but IllegalArgumentException is thrown instead. The system-wide resolver is never called from these new methods.
2. No reverse lookup is performed to resolve a hostname for the supplied address literal - the `InetAddress[46 ]` instances returned by the new `ofLiteral` API has no hostname set.
3. Each `ofLiteral` static method returns addresses of its class only. It gives the ability to check if an IP address literal is of a specific address type. 

### The list of noteworthy changes
- `IPv4-mapped IPv6 address` and `IPv4-compatible IPv6 addresses` require some special handling in the new API to implement all supported IP address types.  
- All address literal parsing code has been moved from `InetAddress.getAllByName` to address type-specific `Inet4Address.parseAddressString` and `Inet6Address.parseAddressString` methods.
- The text with scoped IPv6 addresses architecture draft IETF file has been replaced from `[draft-ietf-ipngwg-scoping-arch-04.txt]` to reference `RFC 4007: IPv6 Scoped Address Architecture`. The "RFC 4007" has been also added as `@ spec` into Inet6Address class-level Javadoc.

### Testing 

`jdk-tier1`, `jdk-tier2`, and `jdk-tier3` test sets show no failure with the changes.

`java/net` JCK tests are failing with new methods added failure (CSR is planned for this change):
```
Added Methods
-------------

java.net.Inet4Address:                  method public static java.net.Inet4Address java.net.Inet4Address.ofLiteral(java.lang.String)
java.net.Inet6Address:                  method public static java.net.InetAddress java.net.Inet6Address.ofLiteral(java.lang.String)
java.net.InetAddress:                   method public static java.net.InetAddress java.net.InetAddress.ofLiteral(java.lang.String)
```

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8312170](https://bugs.openjdk.org/browse/JDK-8312170) to be approved

### Issues
 * [JDK-8272215](https://bugs.openjdk.org/browse/JDK-8272215): Add InetAddress methods for parsing IP address literals (**Enhancement** - P4)
 * [JDK-8312170](https://bugs.openjdk.org/browse/JDK-8312170): Add InetAddress methods for parsing IP address literals (**CSR**)


### Reviewers
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**)
 * [Michael McMahon](https://openjdk.org/census#michaelm) (@Michael-Mc-Mahon - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15775/head:pull/15775` \
`$ git checkout pull/15775`

Update a local copy of the PR: \
`$ git checkout pull/15775` \
`$ git pull https://git.openjdk.org/jdk.git pull/15775/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15775`

View PR using the GUI difftool: \
`$ git pr show -t 15775`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15775.diff">https://git.openjdk.org/jdk/pull/15775.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15775#issuecomment-1722480943)

AlanBateman commented:
At a. high-level, introducing static methods named "ofLiteral" to parse a string as an IPv4 or IPv6 literal address is good. My guess is that the methods will require work on the class descriptions, speak bout parsing the input string, and only mention "blocking" in the context of specifying that they don't attempt reverse name resolution. My guess is that the API additions will require a bit of re-work of the class descriptions, esp. as they are specified in terms of "textual representation".

On Inet4Address.ofLiteral, I can't tell from the method description if it supports all 4 forms that are listed in the class description. It may a surprise that this method parses IPv4-mapped IPv6 addresses. So I think my main comment here is that the method description will need to more completely specify the forms that the input string may take.

Similar comment on Inet6Address.ofLiteral. I can't tell from the API docs if the input needs enclosing square brackets, I can't tell if it supports shortened forms, I can't tell if it's legal or not to include a scope ID. So I think my comment is just that you'll probably go through a few iterations to fully specify the forms that the input string can take.

InetAddress.ofLiteral doesn't need "throws IAE" in the method signature as it's an unchecked exception. I think this method will end up linking to the other two, and specifying that it first attempts to parse the input as an IPv4 address.

My comments here are just about the API for now, I didn't look at the implementation or tests at the time.

msheppar commented:
Looking at the motivation for this addition in the budgd description
The requirement is that of providing a methof to parse of an  string literal address returns and return an InetAddress without necessitating a reverse address lookup, which is what getByName does at the minute.
The additional motivation, appears to be to avoid the  misinterpretation scenario where an invalid address
String literal is considered as a hostname or fqdn resulting in a look by name.

Adding explicit factory methods is one approach, as proposed here. 
An alternative is to provide a string literal validator
which returns true if the string literal address is either an IPv4 or IPv6 address.
The validator can be used by an application in conjunction with the existing API getByName method.
For example, boolean isValidLiteralAddress(String literalAddress)
 InetAddress serverAddress;
 if (InetAddress.isValidLiteralAddress(someLiteralAddress) {
     serverAddress = InetAddress.getByName(someAddress);
 …
 }

This will provide applications with a convenience method to validate inputs and use existing API without side effects.
Also, this approach will reduce the amount of internal restructuring required, should guarantee current behaviour, and provide a convenience application validation method.

But, for some might consider making two method calls  be too much of a burden!

If this is not considered an appropriate option, then a renaming of the proposed methods would be beneficial so as to exhibit the semantics of the operations.
For example, in keeping with the current nomenclature of getByName, getByAddress then getFromLiteral or getByLiteral would be in keeping with current naming scheme.

Alternatively, a straight forward “it does what it says on the tin” name  such as parseLiteral would seem appropriate. I'm not a big fan of these "of" factory methods.

msheppar commented:
on further reflection, wrt method name, an overloaded getByAddress(String literalAddress) method is most appropriate rather then the of methods

ecki commented:
> on further reflection, wrt method name, an overloaded getByAddress(String literalAddress) method is most appropriate

I agree, and it should have both variants, getByAddtess(String literalSddress) throes UHE
GetByAdress(String hostname, String, literalSdddress) throws UHE


In order to produce resolved addresses as well. 

(The UHE is a bit strange but the preexisting method used it for a reason for invalid address arrays, I guess)

dfuch commented:
The main motivation here is to have a method that will create an InetAddress from an IP literal, and throw if the string given as parameters doesn't correspond to an IP literal. By comparison, `getByName` will try to parse the given name as an IP literal, and if it can't parse it as an IP literal, pass it to the resolver. The new method will never invoke the resolver and throw in that case.

msheppar commented:
> The main motivation here is to have a method that will create an InetAddress from an IP literal, and throw if the string given as parameters doesn't correspond to an IP literal. By comparison, `getByName` will try to parse the given name as an IP literal, and if it can't parse it as an IP literal, pass it to the resolver. The new method will never invoke the resolver and throw in that case. I find the name `ofLiteral` to be closer to what the method does than it would be if we provided it as an override to e.g. `getByAddress`, but naming can be discussed.

Well, the getByAddress methods are already providing a variation on the proposed functionality, and as such it makes sense from an OO perspective (polymorphism) to provide an overloaded method to perform the proposed functionality.  getByAddress is doing "what is says on the tin" ... ofLiteral conveys no semantics  and is more cryptic or nuanced

ecki commented on src/java.base/share/classes/java/net/Inet4Address.java:
```diff
@@ -134,6 +136,61 @@ class Inet4Address extends InetAddress {
+     * dotted-decimal form:
+     *
+     *   <blockquote><ul style="list-style-type:none">
+     *   <li>{@code 1.2.3.4}</li>
+     *   <li>{@code 06.07.08.09}</li>
```
>      *     Inet4Address.ofLiteral("06.07.08.09")        ==>    /6.7.8.9

if we ignore octal encoding, we should document it explicitely.

Maybe 

> The loose IPv4 syntax of partial numbers or hex encoding ("10.0.258", "0xA000102") is not supported and causes a parsing error. Octal (zero) prefixes are ignored, i.e. interpreted decimal.

(However I would prefer to reject octal in this case)

dfuch replied:
This method behaves like `InetAddress.getByName` in that regards.

dfuch replied:
I believe accepting leading 0 as an octal indicator would require a new API, yet to be defined. What we're proposing here is a method that behaves like `InetAddress.getByName` would except that it rejects what `InetAddress.getByName` would have identified as a hostname needing resolution. In other words it only accepts IP literals that `InetAddress.getByName` would accept as IP literals, and parses them as `InetAddress.getByName` would parse them, and rejects anything that would cause `InetAddress.getByName` to invoke the resolver.

AlekseiEfimov replied:
> if we ignore octal encoding, we should document it explicitely.

Agree - Javadoc for `Inet4Address.ofLiteral` needs to be updated to cleary mention that only IPv4 literals in decimal encoding are supported, and to state that no other encodings (octal, hexadecimal) are supported.  That will also need to mention that one, two, three and four-part IPv4 address literals are supported to match [Inet4Address format section](https://docs.oracle.com/en/java/javase/20/docs/api/java.base/java/net/Inet4Address.html#format)

> Maybe
>
> > The loose IPv4 syntax of partial numbers or hex encoding ("10.0.258", "0xA000102") is > > not supported and causes a parsing error. Octal (zero) prefixes are ignored, i.e. interpreted decimal.

Thanks for the suggestion. `10.0.258` is supported and [documented](https://docs.oracle.com/en/java/javase/20/docs/api/java.base/java/net/Inet4Address.html#format) 'd.d.d' format. It is not planned to change support for such address.

dfuch commented:
Well the name `ofLiteral` clearly conveys that it only accepts a string that represents an IP literal: I'm not sure what's cryptic about it?

Michael-Mc-Mahon commented:
I'd have a mild preference for the new name rather than overloading the existing `getByAddress`. With IDE auto completion, I think a distinct name is a bit clearer for the developer as to what exactly the method will do. 

I think if it can be specified and implemented that `InetAddress.ofLiteral` calls the sub-type variants (based on string length maybe) then I think it makes sense to specify the parsing details in the sub-classes. If not, then I'm not sure we need the new methods in the sub-classes.

AlekseiEfimov commented:
> At a high-level, introducing static methods named "ofLiteral" or "parse" to parse a string as an IPv4 or IPv6 literal address is good. I think the methods will require work on the class descriptions, speak about parsing the input string, and only mention "blocking" in the context of specifying that they don't attempt reverse name resolution.

Thanks Alan. Based on the intial feedback it is clear that these new methods require more work on documentation. 

> On Inet4Address.ofLiteral, I can't tell from the method description if it supports all 4 forms that are listed in the class description. It may be a surprise that this method parses IPv4-mapped IPv6 addresses. So I think my main comment here is that the method description will need to more completely specify the forms that the input string may take.

I agree that seeing `Inet4Address.ofLiteral` capable of parsing IPv4-mapped IPv6 addresses can look strange here. On contrary, there should be no surprise for `Inet6Address.ofLiteral` to parse such addresses. I will pursue updating its signature to return `InetAddress` and document it as one of the supported forms.

> Similar comment on Inet6Address.ofLiteral. I can't tell from the API docs if the input needs enclosing square brackets, I can't tell if it supports shortened forms, I can't tell if it's legal or not to include a scope ID. 

`Inet6Address.ofLiteral` supports literal enclosed with/without square brackets, supports scope ID, shorten-format. But it is not mentioned in Javadoc. I will work on updating Javadoc to list all these aspects.

AlekseiEfimov commented:
Thanks for all the comments/feedback/suggestions provided. They highlight the fact that this change requires additional work in the following areas:
- Improving new methods Javadoc to mention all supported IP address literal forms.
- Having Inet4Addres.ofLiteral responsible for parsing IPv4-mapped IPv6 addresses looks strange. That will require method signature changes, but will keep javadoc and API less confusing.

In comming days I will work on fixing the above issues first. The methods name can be discussed after that.

AlekseiEfimov commented:
The next iteration contains the following changes:
    - Improvements to Javadoc of the new methods. That includes the addition of all supported IP address literal forms.
    - `Inet4Addres.ofLiteral` is no longer capable of parsing IPv4-mapped IPv6 address literals - it can be done with `Inet6Address.ofLiteral` instead.
    - Added missing `null` checks for literal parameters.
    - Added new test scenarios.

AlekseiEfimov commented:
To restart the conversation around the method name, I've prepared the following table with a list of method names discussed during the previous review iteration:

| Name | Usage Example | Notes |
|-------|--------------|-------|
| `ofLiteral` | `InetAddress.ofLiteral` | Current version. Cleary conveys it takes only an IP literal. Matches its implementation. Good for IDE autocompletion. |
| `parse` | `InetAddress.parse` | The fact that only IP address literals are taken as a parameter is obscured a bit. But still suggests that the InetAddress instance is constructed by parsing some text. Good for IDE.|
| `parseLiteral` | `InetAddress.parseLiteral` | Method name changed to hint that an IP address literal is taken as a parameter. Good for IDE. Matches its implementation.|
| `getByAddress` | `InetAddress.getByAddress` | Overrides existing method in InetAddress, but adds new method to Inet4Address and Inet6Address. Introduces new meaning of `address` term - IP address literal string. In existing API the `address` means a raw byte array.|

My preference is to follow with one of the following two:
`ofLiteral` or `parseLiteral`.

msheppar commented:
the most consistent (with the current API) and object oriented version is the getByAddress(String addr). It is consistent with the existing API. Logically getByAddress(byte[]) and getByAddress(String) are essentially the same. That is to say, a literal string representing an address is essentially the same as an array of bytes representing an address, as a String is an encapsulation of an array of bytes. As such it logically natural to add an overloaded method.
Adding a new style of method affects the logical balance of the InetAddress class, giving a mongrel, or hybrid API appearance.

Ceist agam ort:
Why is it necessary that the subclasses Inet4Address and Inet6Address have an explicit public method for a getByAddress approach? Or have I misconstrued your statement:  "Overrides existing method in InetAddress, but adds new method to Inet4Address and Inet6Address." ?
But this is the case for the current "ofLiteral" solution presented in the PR, in that each of the derived classes Inet4Address and Inet6Address have added new static public methods.

AlekseiEfimov commented:
> That is to say, a literal string representing an address is essentially the same as an array of bytes representing an address

 A literal IP address string can specify the same address in [multiple formats](https://docs.oracle.com/en/java/javase/20/docs/api/java.base/java/net/InetAddress.html#textual-representation-of-ip-addresses-heading), a byte array supports only one - an array of bytes, with length equal to 4 for IPv4 addresses, and to 16 for IPv6 addresses.

> Ceist agam ort:
> Why is it necessary that the subclasses Inet4Address and Inet6Address have an explicit public method for a getByAddress approach?

The idea of this change is to add three methods capable of parsing a string as an IPv4, IPv6 or any IP literal address. The explicit methods needs to be added to all three classes irrelevant to their naming.

> Or have I misconstrued your statement: "Overrides existing method in InetAddress, but adds new method to Inet4Address and Inet6Address." ?
> But this is the case for the current "ofLiteral" solution presented in the PR, in that each of the derived classes Inet4Address and Inet6Address have added new static public methods.

You are right that the statement is a bit vague, yes. What was meant there that overloading `InetAddres` method, and adding new methods to `Inet4Address` and `Inet6Address` would introduce a bit confusing API structure. All three classes for both approaches would look like:

### ofLiteral approach
```
// New methods
InetAddress.ofLiteral(String)
Inet4Address.ofLiteral(String)
Inet6Address.ofLiteral(String)
```

### getByAddress approach
```
// Existing methods with `address` being a byte array
InetAddress.getByAddress(String, byte[])
InetAddress.getByAddress(byte[])

// New methods with `address` being a literal  
InetAddress.getByAddress(String) <-- overloaded method
Inet4Address.getByAddress(String) <-- new and not overloaded method
Inet6Address.getByAddress(String) <-- new and not overloaded method
```

msheppar commented:
thanks for the considered response

> > That is to say, a literal string representing an address is essentially the same as an array of bytes representing an address
> 
> A literal IP address string can specify the same address in [multiple formats](https://docs.oracle.com/en/java/javase/20/docs/api/java.base/java/net/InetAddress.html#textual-representation-of-ip-addresses-heading), a byte array supports only one - an array of bytes, with length equal to 4 for IPv4 addresses, and to 16 for IPv6 addresses.

I think you may have misinterpreted the point I was trying to make (or I didn't make it clearly enough), for example, a 4 byte array representing an IPv4 address is logically the same as a String representing an IPv4 address in dot notation. Thus, overloading getByAddress is a reasonable method to add and exhibits OO characteristics.

> 
> > Ceist agam ort:
> > Why is it necessary that the subclasses Inet4Address and Inet6Address have an explicit public method for a getByAddress approach?
> 
> The idea of this change is to add three methods capable of parsing a string as an IPv4, IPv6 or any IP literal address. The explicit methods need to be added to all three classes irrelevant to their naming.
> 
> > Or have I misconstrued your statement: "Overrides existing method in InetAddress, but adds new method to Inet4Address and Inet6Address." ?
> > But this is the case for the current "ofLiteral" solution presented in the PR, in that each of the derived classes Inet4Address and Inet6Address have added new static public methods.
> 
> You are right that the statement is a bit vague, yes. What was meant there was overloading the `InetAddress` method, and adding new methods to `Inet4Address` and `Inet6Address` would introduce a bit confusing API structure. All three classes for both approaches would look like:
> ### ofLiteral approach
> 
> ```
> // New methods
> InetAddress.ofLiteral(String)
> Inet4Address.ofLiteral(String)
> Inet6Address.ofLiteral(String)
> ```
> 
> ### getByAddress approach
> 
> ```
> // Existing methods with `address` being a byte array
> InetAddress.getByAddress(String, byte[])
> InetAddress.getByAddress(byte[])
> 
> // New methods with `address` being a literal  
> InetAddress.getByAddress(String) <-- overloaded method
> Inet4Address.getByAddress(String) <-- new and not overloaded method
> Inet6Address.getByAddress(String) <-- new and not overloaded method
> ```

There's no need to add these methods,
Inet4Address.getByAddress(String) <-- new and not overloaded method
Inet6Address.getByAddress(String) <-- new and not overloaded method

 to the derived classes. The functionality can be fully provided by the overloaded (static) getByAddress in the base InetAddress class

AlanBateman commented:
> The functionality can be fully provided by the overloaded (static) getByAddress in the base InetAddress class

I don't think I agree with this. The existing 2-arg getByAddress method takes a string that is a host name or string representation of an IP address. It can be argued that it was a mistake to allow it accept IP literal addresses but it can't be changed. Same thing with the UHE thrown by both getByAddress methods but that can't be changed either.

The new factory methods have a very clear purpose to create an InetAddress from the string representation of an IP literal address. The proposed name ofLiteral is readable/clear and consistent with the naming of other static factory methods added to the APIs in recent years.








The goal here is to add methods to parse IP address literals.

msheppar commented:
> > The functionality can be fully provided by the overloaded (static) getByAddress in the base InetAddress class
> 
> I don't think I agree with this. The existing 2-arg getByAddress method takes a string that is a host name or string representation of an IP address. It can be argued that it was a mistake to allow it accept IP literal addresses but it can't be changed. Same thing with the UHE thrown by both getByAddress methods but that can't be changed either.
> 

The sentence should have read "The functionality can be fully provided by adding an overloaded (static) getByAddress in the base InetAddress class, only."

I wasn't suggesting that the current API was providing the requested functionality.
you may have misinterpreted what I'm saying. 
The requested functionality can be provided by an overloaded getByAddress method to the InetAddress , and that is sufficient. As such, there is no need for additional method in the Inet4Address or Inet6Address

> The new factory methods have a very clear purpose to create an InetAddress from the string representation of an IP literal address. The proposed name ofLiteral is readable/clear and consistent with the naming of other static factory methods added to the APIs in recent years.

For me, it is not consistent with the well esatblished InetAddress API (getByName, getByAddress), in which case the ofLiteral style looks like a bit of a wart.  Adding an overloaded getByAddress method is a logical OO design approach.  You seem to be intent on adding this "ofXXX" style methods. So we'll just have to agree to disagree on this.

AlanBateman commented:
> The sentence should have read "The functionality can be fully provided by adding an overloaded (static) getByAddress in the base InetAddress class, only."
> 
> I wasn't suggesting that the current API was providing the requested functionality. you may have misinterpreted what I'm saying. The requested functionality can be provided by an overloaded getByAddress method to the InetAddress , and that is sufficient. As such, there is no need for additional method in the Inet4Address or Inet6Address

The existing naming in InetAddress comes with baggage. The getByXXX method suggests lookup, they throw UHE. The getByAddress methods take a byte[], the input is bytes. The String parameter for the 2-arg getByAddress is intended to be a host name. I don't think the new static factory methods should be burdened with all this baggage so move to a methods that are clear/obvious/readable seems a much better route. We can bikeshed over specific name but the ofXXX convention is something that many areas of the platform has adopted in recent years.

jaikiran commented:
Hello Aleksei, it's good to see the addition of these new APIs. On the naming front, I like the proposal of using either `ofLiteral` or `parseLiteral`.

The new APIs are proposing to throw an `IllegalArgumentException` if the passed literal is not an IP address literal. That seems logical. However, the existing `getByAddress(byte[] ...)` API which accepts a byte form of the raw IP address, currently throws `UnknownHostException` in the case where the IP address isn't of a valid length. Should we consider throwing `UnknownHostException` from these new APIs too? I prefer `IllegalArgumentException` in these new APIs, but `UnknownHostException` might be something that would need to be considered?

The proposed changes also have text which talks about "ambiguous" address literals. The term ambiguous seems new in the InetAddress (and sub-classes). The package-info.java too doesn't have previous references to it. Should we add some details on what addresses are considered ambiguous (and how/if such addresses are treated by existing APIs)?

jaikiran commented on src/java.base/share/classes/sun/net/util/IPAddressUtil.java:
```diff
@@ -134,22 +134,39 @@ public static byte[] textToNumericFormatV4(String src)
      * If string can't be parsed by following IETF IPv4 address string literals
      * formatting style rules (default one), but can be parsed by following BSD formatting
```
I realize that this is an existing javadoc comment in an internal utility class (which is used by public InetAddress API implementation); do you think we should add a more formal reference to the BSD formatting of IP address, like we currently do for other RFCs and such in the class level javadoc of `InetAddress`?

AlanBateman commented:
> The new APIs are proposing to throw an `IllegalArgumentException` if the passed literal is not an IP address literal. That seems logical. However, the existing `getByAddress(byte[] ...)` API which accepts a byte form of the raw IP address, currently throws `UnknownHostException` in the case where the IP address isn't of a valid length. Should we consider throwing `UnknownHostException` from these new APIs too? I prefer `IllegalArgumentException` in these new APIs, but `UnknownHostException` might be something that would need to be considered?

UHE is an IOException so would be problematic for a number of reasons. IAE is the right exception for bad input here.

jaikiran commented on src/java.base/share/classes/java/net/Inet4Address.java:
```diff
@@ -134,6 +137,64 @@ class Inet4Address extends InetAddress {
+
+    /**
+     * Parses string with an IPv4 address literal.
+     * If string doesn't contain a valid literal - null is returned.
```
This sentence is a bit confusing, given the implementation details of this method. I think `null` is only returned if `throwIAE` is false?

jaikiran commented on src/java.base/share/classes/sun/net/util/IPAddressUtil.java:
```diff
@@ -134,22 +134,39 @@ public static byte[] textToNumericFormatV4(String src)
      *
      * @param src input string
+     * @param throwIAE {@code "true"} - throw {@code IllegalArgumentException} when cannot be parsed as IPv4 address string;
+     *                 {@code "false"} - throw {@code "IllegalArgumentException"} only when IPv4 address string is ambiguous.
```
Are these double quotes around `true`, `false`, `null` and `IllegalArgumentException` intentional? It seems odd to have those double quotes when (rightly) using `{@code}`.

AlekseiEfimov replied:
Nope, there are not intentional - a leftover from the fix prototyping stage. Will remove them.

AlekseiEfimov commented:
> Hello Aleksei, it's good to see the addition of these new APIs. On the naming front, I like the 

Hello Jaikiran

> The proposed changes also have text which talks about "ambiguous" address literals. The term ambiguous seems new in the InetAddress (and sub-classes). The package-info.java too doesn't have previous references to it.

What ambiguous IPv4 address literals mean in the code and in [the release notes](https://www.oracle.com/java/technologies/javase/18all-relnotes.html#JDK-8277608): it is a set of addresses which can't be parsed in dotted-decimal forms listed in the Inet4Address#format class-level javadoc section, but can be parsed in BSD forms (decimal, octal, or hexadecimal) defined in [`inet_addr` of IEEE Std 1003.1-2017](https://pubs.opengroup.org/onlinepubs/9699919799/functions/inet_addr.html).

> Should we add some details on what addresses are considered ambiguous (and how/if such addresses are treated by existing APIs)?

The release note states that the `InetAddress` accepts IPv4 address literals in decimal quad notation. The new `ofLiteral` API also states that only dotted-decimal forms are supported, `ambigous` term is not mentioned in any public javadoc, and given that the related system property is a jdk-specific system property I would like to avoid introducing this term as part of this PR. 

> do you think we should add a more formal reference to the BSD formatting of IP address

I think a more formal reference to BSD forms might be needed if they are recognizable by InetAddress APIs. As of today, this is not the case.

Michael-Mc-Mahon commented on src/java.base/share/classes/java/net/Inet4Address.java:
```diff
@@ -134,6 +137,64 @@ class Inet4Address extends InetAddress {
+     *  // the constructed address bytes without any rearrangement
+     *  Inet4Address.ofLiteral("02130706689") ==> /127.0.1.1
+     * }
+     * <p>If the provided address literal cannot represent a valid IPv4 address an
```
I think this behavior is the same as existing methods, except that only the decimal format is supported? I wonder is it really necessary to describe it here, as it it would be quite an arcane usage, and the section in the class doc "Textual representation if IP addresses" already partially covers it. The snippet example with the leading zero seems to be implying that octal format is not supported, but without really saying it out loud. 

Maybe the apidoc for the method could just say that the decimal formats are the only ones supported, but otherwise refer to the class docs for how the different variants work ? And maybe move the snippet there as well?

Michael-Mc-Mahon commented on src/java.base/share/classes/java/net/Inet6Address.java:
```diff
@@ -477,6 +483,146 @@ public static Inet6Address getByAddress(String host, byte[] addr,
+     * @throws IllegalArgumentException if the {@code ipv6AddressLiteral} cannot be
+     *         parsed as an IPv6 address literal.
+     * @throws NullPointerException if the {@code ipv6AddressLiteral} is {@code null}.
+     */
```
Similar comment to Inet4Address. The snippet should be incorporated into the general class docs section "Textual representation of IP addresses" and the apidoc here should limit itself to any limitations that only apply to this method imo.

AlekseiEfimov replied:
That is a good point Michael. The snippets were moved to class-level javadocs, and were modified a bit to limit the number of examples, and other small changes. Pushed as [e95858e](https://github.com/openjdk/jdk/pull/15775/commits/e95858ebb6039e9b5bbbca2a4730fcbed78747ca).

dfuch commented on src/java.base/share/classes/java/net/Inet4Address.java:
```diff
@@ -137,51 +138,61 @@ class Inet4Address extends InetAddress {
+     * Parses string as an IPv4 address literal.
+     * If string contains a non-parsable literal and {@code throwIAE} is set to {@code false}
+     * - {@code null} is returned.
+     * If string contains a non-parsable literal and {@code throwIAE} is set to {@code true}
+     * - {@code IllegalArgumentException} is thrown.
```
```suggestion
     * If the given {@code addressLiteral} string cannot be parsed as an IPv4 address literal
     * and {@code throwIAE} is {@code true}, an {@code IllegalArgumentException} is thrown.
```

dfuch commented on src/java.base/share/classes/java/net/Inet4Address.java:
```diff
@@ -134,6 +160,55 @@ class Inet4Address extends InetAddress {
+     * and {@code throwIAE} is {@code true}, an {@code IllegalArgumentException} is thrown.
+     * If string contains an {@linkplain IPAddressUtil#validateNumericFormatV4(String, boolean)
+     * ambiguous literal} - {@code IllegalArgumentException} is thrown irrelevant to
+     * {@code throwIAE} value.
```
```suggestion
     * {@code throwIAE} value.
     *
     * @apiNote
     * The given {@code addressLiteral} string is considered ambiguous if it cannot be parsed as
     *  a valid IPv4 address literal using decimal notation, but could be
     * interpreted as an IPv4 address in some other representation (octal, hexadecimal, or mixed).
```

AlekseiEfimov replied:
Thanks for the suggestions Daniel. They look reasonable to me, and improve the readability. Taken them aboard in 4bd63f29386a515d63879412f80d958046e6c8c2.

dfuch commented on src/java.base/share/classes/java/net/Inet4Address.java:
```diff
@@ -134,6 +160,55 @@ class Inet4Address extends InetAddress {
+     *                 an IPv4 address literal.
+     * @return {@code Inet4Address} object constructed from the address literal;
+     *         or {@code null} if the literal cannot be parsed as an IPv4 address
+     * @throws IllegalArgumentException if the given {@code addressLiteral} string
+     * cannot be parsed as an IPv4 address literal and {@code throwIAE} is {@code true}.
+     * An {@code IllegalArgumentException} is also thrown regardless of the value of
+     * {@code throwIAE} if the given {@code addressLiteral} string is ambiguous, that is,
+     * it cannot be parsed as a valid IPv4 address literal using decimal notation but could be
+     * interpreted as an IPv4 address in some other representation (octal, hexadecimal, or mixed).
```
```suggestion
     * @throws IllegalArgumentException if the given {@code addressLiteral} string
     * cannot be parsed as an IPv4 address literal and {@code throwIAE} is {@code true},
     * or if it is considered ambiguous, regardless of the value of the value of
     * {@code throwIAE}.
```

dfuch reviewed (commented):
Sorry for the back and forth suggestions - I hadn't noticed that "ambiguous" was already introduced earlier in the method description. I'd suggest the following refactoring to make it easier to read.

dfuch commented on src/java.base/share/classes/java/net/Inet4Address.java:
```diff
@@ -134,6 +160,55 @@ class Inet4Address extends InetAddress {
+     * and {@code throwIAE} is {@code false}, {@code null} is returned.
+     * If the given {@code addressLiteral} string cannot be parsed as an IPv4 address literal
+     * and {@code throwIAE} is {@code true}, an {@code IllegalArgumentException} is thrown.
+     * If string contains an {@linkplain IPAddressUtil#validateNumericFormatV4(String, boolean)
+     * ambiguous literal} - {@code IllegalArgumentException} is thrown irrelevant to
```
```suggestion
     * Otherwise, if it can be considered as {@linkplain IPAddressUtil#validateNumericFormatV4(String, boolean)
     * an ambiguous literal} - {@code IllegalArgumentException} is thrown irrelevant to
```

Michael-Mc-Mahon commented on src/java.base/share/classes/java/net/Inet6Address.java:
```diff
@@ -477,6 +495,128 @@ public static Inet6Address getByAddress(String host, byte[] addr,
 
+    /**
+     * Creates an {@code InetAddress} based on the provided {@linkplain
+     * Inet6Address##format textual representations} of an IPv6 address.
```
"textual representations" -> "textual representation"

AlekseiEfimov replied:
Fixed in af87fab27148daa982ab1997a4346e8c29be7604

Michael-Mc-Mahon commented on src/java.base/share/classes/java/net/Inet4Address.java:
```diff
@@ -134,6 +160,57 @@ class Inet4Address extends InetAddress {
+    /**
+     * Creates an {@code Inet4Address} based on the provided {@linkplain
+     * Inet4Address##format textual representations} of an IPv4 address.
+     * <p> If the provided IPv4 address literal cannot represent a {@linkplain
```
"textual representations" -> "textual representation"

AlekseiEfimov replied:
Thanks Michael, fixed in af87fab27148daa982ab1997a4346e8c29be7604

dfuch reviewed (approved):
LGTM

Michael-Mc-Mahon reviewed (approved):
LGTM

AlekseiEfimov commented:
/integrate
